---
title: "DGE_analysis_estimateLF"
author: ""
date: "2024"
output:
  word_document: default
  pdf_document: default
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Model:

Gene expression ~ treatment + covariates

```{r load library, echo=FALSE, message=FALSE, warning=FALSE}
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(tidyverse)
library(limma)
library(parallel)
library(FALCO)
library(CorrConf)
library(BCconf)
#devtools::install_github(repo="chrismckennan/BCconf/BCconf")
```


```{r read in data, echo=FALSE}
# read in clean/corrected normalized gene expression data

#load("output/dge_corrected.RData")

ge_full <-  read.csv("output/limma-trend_GE2.csv")
ge <-  ge_full[4:114]
#dge <- DGEList(counts = ge[, 4:114], genes = ge[, 1:3])

# read in all covariates
covs_all <- read.csv("output/covs_gds.csv")

# extract covariates of interest
covs <- covs_all %>% 
  select(1, 2, 4:6, 8:14) %>% 
  column_to_rownames(var="RNA_sampleID") %>% 
  mutate_at(c("sex", "treatment"), as.factor)

#all.equal(rownames(covs), colnames(ge))
```

#### Estimate latent factors
```{r echo=FALSE}
############# Covariates when subject is treated as a random effect
# creating a design matrix
ge <- as.matrix(ge)

design <- model.matrix(~ treatment + covariates, covs)
subj <- model.matrix(~as.factor(covs$Subject_ID) - 1)
randomE <- list(diag(ncol(ge)), subj%*%t(subj))    #B list when subject included as a random effect

set.seed(1)
falco_random <- FALCO::CBCV_plus(Y = ge, Cov = design, maxK = 30, B = randomE, nFolds = 3, simpleDelta = T)

factor <- falco_random$K.hat
correct <- Correction(ge, design, 2, factor, method="ols", shrink.Sigma=F)
cov_random_ols <- correct$Cov.total

write.table(cov_random_ols, "output/cov_random_LF.txt")
```

Differential gene expression analysis

**Treatment effect (FDR 0.05)**
```{r echo=FALSE}
# set up contrasts for pairwise comparisons 
#contr.matrix <- makeContrasts(OsP = treatment1, levels = colnames(design))
 
# random effect
random_effect <- duplicateCorrelation(ge, cov_random_ols, block = covs$Subject_ID)
vfit <- lmFit(ge, cov_random_ols, block = covs$Subject_ID, correlation=random_effect$consensus.correlation)
#vfit <- contrasts.fit(vfit, contrasts=contr.matrix)
efit <- eBayes(vfit)
#summary(decideTests(efit, adjust.method = "BH", p.value = 0.05))[, 1:8]

a = topTable(efit, coef = colnames(cov_random_ols)[2], adjust.method="fdr", number=dim(ge)[1])
b = merge(ge_full[, 1:3], a, by = 0) %>%  # merge by row names (by=0 or by="row.names") all.x=TRUE. This is known as a LEFT JOIN
  arrange(P.Value) %>% 
  column_to_rownames(var="Row.names")

# subset with significant p-values
b_sig <- subset(b, adj.P.Val < 0.05)
#plotSA(efit, main="Final model: Mean-variance trend")
```

#### Data visualization
Volcano plot can be use to visualize the genes that are differentially expressed at a given fold-change and p-value
```{r echo=FALSE}
hlcol <- which(b$adj.P.Val<= 0.05)

ggplot(b, aes(x=logFC, y=-log10(adj.P.Val))) +
    geom_point(aes(color="grey", size=0.7, alpha=0.4)) +
    ggtitle('Volcano Plot') +
    labs(y=expression('-Log'[10]*' P'[adj]), x=expression('Log'[2]*' fold change')) +
    theme_minimal() +
    theme(legend.position="none", plot.title = element_text(size = rel(1.5), hjust = 0.5))+
    geom_text_repel(data=b[hlcol,], aes(x = logFC, y = -log10(adj.P.Val), label=symbol), size=2)+
    geom_point(data=b[hlcol,], aes(x = logFC, y = -log10(adj.P.Val), color="red", size=0.7, alpha=0.4))+
    scale_color_manual(values=c('gray','red'))
```

```{r echo=FALSE, warning=FALSE}
znf <- ge_full %>% 
  filter(symbol == "gene_name") %>% 
  pivot_longer(cols = CO_CW_001:CO_CW_123,
               names_to = "Samples",
               values_to = "gene_name") %>% 
  left_join(covs_all[, c(1, 13)], join_by(Samples == RNA_sampleID)) %>% 
  select(-(1:3)) %>% 
  mutate_at("treatment", as.factor)

ggplot(znf, aes(treatment, ZNF154)) +
  geom_boxplot() +
  labs(subtitle = "gene_name", y="logCPM", x="P_T") +
  scale_y_continuous(limits = c(-4, 3))

