---
title: "OMG for Visit 1 on Aug. 22"
author: ''
date: "08/22/2024"
output:
  word_document: default
  pdf_document: default
  df_print: paged
  html_document: null
---

```{css, echo=FALSE}
h3{text-align: center; 
  color:#6495ED}
   
h4{text-align: right;
  color:seagreen;}

h5{text-align:left;
  color:PaleVioletRed;}
  
p{color: black;}
```
**Outline **

-Filter out lowly expressed genes for visit 1

-Regress out _RIN_ and _pct_ to get residuals for further analysis

-Use limma to investigate correlations between gene expression (residuals) and OMG within each time (with appropriate covariates and/or latent factors)

```{r echo=FALSE, warning=FALSE, message=FALSE}
library(edgeR)
library(ggplot2)
library(tidyverse)
library(reshape2)
library(variancePartition)
library(ggpubr)
library(ggrepel)
library(GWASTools)
```

To better know our data, summary statistics of numerical variables and sample size for categorical variables were shown as below.

Summary statistics for numerical variables shown below
```{r echo=FALSE, warning=FALSE, message=FALSE}
# Read in meta data

meta <- read.csv("output/OMG_meta.csv") %>% 
  drop_na(OMG) %>% 
  mutate(OMG_factor=cut(OMG, breaks=c(0, 5, 7.5, 13), labels = c("Low", "Medium", "High")), .after=OMG) %>% 
  mutate_at(c('Visit', 'sequence_pool', 'sex'), factor)       # change categorical variables into factor

meta_v1 <- meta %>% 
  filter(Visit=='v1')

summary(meta_v1[, c(3, 4, 6:10, 12, 16)])
```

Sample size for each OMG group in Visit 1 as shown below.
```{r echo=FALSE}
table(meta_v1$OMG_factor)
```

Sample size for infant sex in Visit 1 as shown below.
```{r echo=FALSE}
table(meta_v1$infant_sex)
```

**Assess correlation between all pairs of variables (Visit 1)**

```{r eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}
form <- ~ RIN + RNA_conc + sequence_pool + pct + variables_of_interest

# Compute Canonical Correlation Analysis (CCA)
# between all pairs of variables.  # returns absolute correlation value
C <- canCorPairs(form, meta_v1)

par(mar=c(1,1,1,1))
# Plot correlation matrix between all pairs of variables

plotCorrMatrix(C,  dendrogram = "none", cexRow=1.2, cexCol=1.2)
```

**Fig. 2 Plot of correlation between all pairs of variables**


### Removing lowly expressed genes (Visit 1)

Before filtering, I first annotated genes using the gene reference (basic gene annotation for GRCh38.p14 downloaded from GENCODE website), then filtered by Class (keeping "protein_coding"), and Chromosomes (removing "chrY" and "chrM"). At last, we have __19,878__ genes left.

__Keeping genes that have a count per million (CPM) > 1 in at least 25% of samples__

__11,544__ genes were left after filtering lowly expressed genes.

```{r echo=FALSE, message=FALSE, warning=FALSE}
# https://bioinformatics-core-shared-training.github.io/RNAseq-R/rna-seq-preprocessing.nb.html

# Read in reads counts
load(file="output/OMG_corrected_anno.RData")
# keep samples' expression only in meta data
omg_corrected_anno <- omg_corrected_anno[, 4:126] %>% 
  select(meta_v1$RNA_sampleID) %>% 
  add_column(omg_corrected_anno[, 1:3], .before=1)
  
# Remove first column from omg_corrected_anno
omg <- omg_corrected_anno %>% 
  column_to_rownames(var = "Geneid") %>% 
  select(-c(1, 2))

# confirm the order of sample in meta_v1 and omg same
# sum(meta_v1$RNA_sampleID == colnames(omg))

omg_dge <- DGEList(omg_corrected_anno)    # library size added
#omg_dge <- calcNormFactors(omg_dge)       # normalization factor added
```


```{r echo=FALSE, warning=FALSE}
keepGenes <- rowSums(cpm(omg)>1) >= (0.25*ncol(omg))

omg_filter <- omg_dge[keepGenes,, keep.lib.sizes=FALSE]   ## update DGEList library size

# store gene symbol  
gene_symbol <-  omg_corrected_anno[, c(1:3)] %>% 
  filter(Geneid %in% omg_filter$genes$Geneid)

#  sum(gene_symbol$Geneid==omg_filter$genes$Geneid)

beforeFiltering_plot <- omg_dge %>% 
  cpm(log = TRUE) %>% 
  melt %>% 
  dplyr::filter(is.finite(value)) %>% 
  ggplot(aes(x = value, colour = Var2)) +
  geom_density() + 
  guides(colour = FALSE) +
  ggtitle("A. Before filtering", subtitle = paste0(nrow(omg), " genes")) +
  labs(x = "logCPM", y = "Density") 

afterFiltering_plot <- omg_dge %>% 
  cpm(log = TRUE) %>% 
  magrittr::extract(keepGenes,) %>%
  melt %>% 
  dplyr::filter(is.finite(value)) %>% 
  ggplot(aes(x = value, colour = Var2)) +
  geom_density() + 
  guides(colour = FALSE) +
  ggtitle("B. After filtering", subtitle = paste0(table(keepGenes)[[2]], " genes"))+
  labs(x = "logCPM", y = "Density")+
  ylim(0, 0.25)

cowplot::plot_grid(beforeFiltering_plot, afterFiltering_plot)
```

__Fig. 3 The density of log-CPM values for raw pre-filtered data (A) and post-filtered data (B) shown for each sample__

Plot A shows lots of lowly expressed genes before filtering. 

Plot B shows peaks of lowly expressed genes were removed after filtering.


##### Normalizing gene expression distributions (Visit 1)

Using the method of **trimmed mean of M-values (TMM)** method to normalize and then remove batch effects. 

TMM normalizes the library sizes to produce effective library sizes. 

CPM values are counts normalized by the effective library sizes.

```{r echo=FALSE, warning=FALSE, fig.width=8, fig.height=6}
library(RColorBrewer)
nsamples <- ncol(omg)
col <- rep(brewer.pal(nsamples, "Paired"), times=9)

# barplot(omg_dge$samples$lib.size, names=colnames(omg_dge))
# Get log2 counts per million
logcounts <- cpm(omg_filter, log = TRUE)

boxplot(logcounts, main="A. Unnormalized data", ylab="log-CPM",col=col, cex.axis=0.6, las=2, cex=0.45)
abline(h=median(logcounts),col="blue")

tmm <- calcNormFactors(omg_filter, method="TMM")    # update normalization factor
lcpm <- cpm(tmm, log=TRUE, prior.count=3)
boxplot(lcpm, col=col, main="", cex.axis=0.6, las=2, cex=0.45)
title(main="B. Normalized data",ylab="Log-CPM")
```

__Fig. 4 Boxplots of log-CPM values showing expression distributions for unnormalised data (A) and normalised data (B) for each sample__

Distributions before normalization are noticeably different. Blue line indicates the median of log-CPM.

Distributions after normalization are similar. 


### PCA

```{r echo=FALSE}
#Make sure the samples in the gene expression data and batch information data in the same sequence.

# ge: TMM-normalized, voom transformed data; rows=genes, columns=samples
# bat: covariate file: batch (make sure columns are as.numeric or as.factor. rows: samples (same order as gene expression (x) columns, columns are batch info)

ge <- lcpm
bat <- meta_v1[, -c(1:2, 11, 13)]
#all.equal(colnames(ge), meta_v1$RNA_sampleID)

# even though voom() function is to compute precision weights for the downstream differential expression analysis, it can't use together with removeBatchEffect()
# vm <- voom(omg_filter, plot=TRUE)
# https://support.bioconductor.org/p/9158852/
```


**Before batch effects removed**

```{r echo=FALSE, warning=FALSE, message=FALSE, cols.print = n}
#Step 1: Find original PCs for gene expression and proportion of variance as follows:
PCs <-  paste0("PC", 1:10) 
PCA1 <- prcomp(t(ge), scale=FALSE, center=TRUE)         # run PCA working on matrix and t() returning a matrix
pca_var1 <- summary(PCA1)$importance
pca_var1_output <- round(pca_var1[, 1:10], 3)*100       # convert proportion of variance explained by PCs to percent
# output
#write.csv(pca_var1_output, "output/pca_var1.csv")

#Step 2: PCs before Batch Effects removed to see if any batch effects are correlated with PCs

bat_pca1 <- matrix(ncol=ncol(bat), nrow=10)            # matrix to store results
rownames(bat_pca1) <- PCs
colnames(bat_pca1) <- colnames(bat)

for(j in 1:ncol(bat_pca1)){                            # look at association between PCj ~ batch i
  for(i in 1:length(PCs)){
    data1= lm(PCA1$x[, i] ~ bat[, j])
    bat_pca1[i,j]=anova(data1)$'Pr(>F)'[1] 
  }
}
pca_cvar1 <- rbind(pca_var1_output, t(bat_pca1))[-c(1,3), ] %>% 
  round(3)  #
rownames(pca_cvar1)[rownames(pca_cvar1)=="Proportion of Variance"] <- "Proportion of Variance (%)"
```


![](output/pcav1_cvar1.png){width=100%}

__Table 1 PCs table before batch effects removed__ 


**Removing two batch effects**

Use removeBatchEffect(), a function from limma, to remove _RIN_ and _pct_ while protecting _OMG_factor_. 

_OMG_factor_ is the variable of our interest, which correlates with *pct*, as shown in Fig.1. That's why it should be protected here. 

```{r echo=FALSE}
# regress out technical effects
design <- model.matrix(~ OMG_factor + RIN + pct, data=bat)
MA_design <- design[, 1:3]          # design matrix relating to treatment conditions to be preserved
batch_design <- design[, -(1:3)]    # covariates -- matrix or vector of numeric covariates to be adjusted for
 
ge2 <- removeBatchEffect(ge, covariates = batch_design, design=MA_design)

PCA2 <- prcomp(t(ge2), scale=FALSE, center=TRUE)     # run PCA na.action=na.omit
        
pca_var2 <- summary(PCA2)$importance
pca_var2_output <- round(pca_var2[, 1:10], 3)*100 

bat_pca2 <- matrix(ncol=ncol(bat), nrow=10) # matrix to store results
rownames(bat_pca2) <- PCs
colnames(bat_pca2) <- colnames(bat)

for(j in 1:ncol(bat)){ # look at association between PCj ~ batch i
  for(i in 1:length(PCs)){
    data1= lm(PCA2$x[,i]~bat[,j])
    bat_pca2[i,j]=anova(data1)$'Pr(>F)'[1] 
  }
}

pca_cvar2 <- rbind(pca_var2_output,t(bat_pca2))[-c(1,3), ] %>% 
  round(3)
rownames(pca_cvar2)[rownames(pca_cvar2)=="Proportion of Variance"] <- "Proportion of Variance (%)"
```
![](output/pcav1_cvar2.png){width=100%}

__Table 2 PCs table after *RIN* and *pct* removed__ 

### Differential expression analysis
Model: *gene_expression ~ OMG_factor + covariates*

```{r echo=FALSE}
design <- model.matrix(~0 + OMG_factor + covariates, meta_v1)

# set up contrasts for pairwise comparisons 
contr.matrix <- makeContrasts(LvsM = OMG_factorMedium-OMG_factorLow, 
                              MvsH = OMG_factorHigh-OMG_factorMedium,
                              HvsL = OMG_factorHigh-OMG_factorLow,
                              levels = colnames(design))

# Fitting linear models in limma
fit <- lmFit(ge2, design)   # use residuals -- normalized count regress out 2 batch effects
tmp <- contrasts.fit(fit, contrasts=contr.matrix)  # test OMG_factor coefficient
efit <- eBayes(tmp)
#plotSA(efit, main="Final model: Mean-variance trend")
```

Pairwise comparisons shown as below,

LvsM = OMG_Medium-OMG_Low

MvsH = OMG_High-OMG_Medium

HvsL = OMG_High-OMG_Low

**Examining the number of DE genes (FDR < 0.05)**

```{r echo=FALSE}
bh_fdr_cutoff <- 0.05
summary(decideTests(efit, adjust.method = "BH", p.value = bh_fdr_cutoff))
```

__Table 3 Number of DE genes for each pairwise comparison__

**Examining individual DE genes from top to bottom**

```{r echo=FALSE}
ML <-  topTable(efit, coef = 1, genelist = gene_symbol, sort.by = "P", n = Inf)
HM <-  topTable(efit, coef = 2, genelist = gene_symbol, sort.by = "P", n = Inf)
HL <-  topTable(efit, coef = 3, genelist = gene_symbol, sort.by = "P", n = Inf) %>% 
   remove_rownames() %>% 
   mutate_at(c(4:6, 8, 9), round, 3)

# subset with significant p-values
#b_sig <- subset(ML, adj.P.Val < 0.05)
head(HL[, -c(1, 3)], n=3)
```

__Table 4 Summary statistics of Top 3 genes__

**Visualization of differential expression results**

```{r eval=FALSE, include=FALSE}
qqPlot(ML$P.Value, main="OMG_Medium vs OMG_Low")

# Inflation (produced by batch effects or uncontrolled variables) is normally tested by using QQ-plot
```


```{r echo=FALSE, message=FALSE, warning=FALSE}
a <- HL
hlcol <- which(a$adj.P.Val<= 0.1)

par(mgp=c(3,0.3,0), mar=c(5,5,5,6)+0.1) #mpg 3rd value moves ticks either close to or far from axis line itself

plot.new()
plot.window(xlim=range(a$logFC), ylim=range(-log10(a$P.Value)))
points(a$logFC, -log10(a$P.Value), col="gray", pch=19)

axis(1)
axis(2, cex.axis = 0.9)
box()

plot.window(xlim=range(a$logFC), ylim=range(-log10(a$adj.P.Val)))
points(a[hlcol,]$logFC, -log10(a[hlcol,]$adj.P.Val), col="red", pch=19, cex = 0.5)
axis(4, cex.axis = 0.9)
abline(h=1.3)
text(1.5,1.3, "FDR 0.05", adj = c(0, -.1), cex=0.8)
abline(h=1)
text(1.5,1, "FDR 0.1", adj = c(0, -.1), cex=0.8)

p=par("usr")
title("Volcano Plot for OMG_High vs OMG_Low", adj=0)
#text(p[1]+0.5, mean(p[3:4]), "-log10(P.Val)", srt = 90, col="red", pos=2)
#text(p[3]+10, mean(p[3:4]), "-log10(adj.P.Val)", srt = 90, col="limegreen", pos=4, adj=0.5)
mtext("-log10(adjusted_P)", side = 4, las=3, line=1.5)
mtext("-log10(observed_P)", side = 2, las=3, line=1.5)
mtext("logFC", side = 1, las=1, line=1.5)
```

**Fig. 5 Volcano plot for OMG_High vs OMG_Low**

Left Y axis shows -log10 of observed p-value while right one shows -log10 of adjusted p-value. X axis indicates log fold change. Two black lines show FDR 0.1 and 0.05. Red dot is the DE gene between high and low OMG groups based on adjusted p-value.

```{r echo=FALSE}
hlcol <- which(a$adj.P.Val<= 0.05)

ggplot(a, aes(x=logFC, y=-log10(adj.P.Val))) +
    geom_point(aes(color="grey", alpha=0.4)) +
    ggtitle('Volcano Plot for OMG_High vs OMG_Low') +
    labs(y=expression('-Log'[10]*' P'[adj]), x=expression('Log'[2]*' fold change')) +
    theme_minimal() +
    theme(legend.position="none", plot.title = element_text(size = rel(1.5), hjust = 0.5))+
    geom_text_repel(data=a[hlcol,], aes(x = logFC, y = -log10(adj.P.Val), label=GeneSymbol))+
    geom_point(data=a[hlcol,], aes(x = logFC, y = -log10(adj.P.Val), color="red", alpha=0.4))+
    scale_color_manual(values=c('gray','red'))
```


**Fig. 6 Volcano plot for OMG_High vs OMG_Low**


```{r echo=FALSE, message=FALSE, warning=FALSE}
gene1 <- cbind(gene_symbol, omg_filter$counts) %>% 
  filter(GeneSymbol == "gene_name") %>% 
  pivot_longer(cols = CO_CW_011:CO_CW_120,
               names_to = "Samples",
               values_to = "gene_name") %>% 
  right_join(meta_v1, join_by(Samples == RNA_sampleID)) %>% 
  select(4, 5, 18) 

ggplot(gene1, aes(OMG_factor, gene_name)) +
  geom_boxplot() +
  labs(subtitle = "Boxplot of DE gene (gene_name)", y="gene1(raw counts)", x="OMG_factor") 

```

**Fig. 7 Box plot of DE gene**


